modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java

```diff
@@ -122,8 +122,34 @@
         Netty4Utils.setup();
     }
 
+    private static final ByteSizeValue MTU_ETHERNET = new ByteSizeValue(1500);
+
+    private static final String SETTING_KEY_HTTP_NETTY_MAX_COMPOSITE_BUFFER_COMPONENTS = "http.netty.max_composite_buffer_components";
+
     public static Setting<Integer> SETTING_HTTP_NETTY_MAX_COMPOSITE_BUFFER_COMPONENTS =
-        Setting.intSetting("http.netty.max_composite_buffer_components", -1, Property.NodeScope);
+        new Setting<>(SETTING_KEY_HTTP_NETTY_MAX_COMPOSITE_BUFFER_COMPONENTS, (s) -> {
+            ByteSizeValue maxContentLength = SETTING_HTTP_MAX_CONTENT_LENGTH.get(s);
+            // Netty accumulates buffers containing data from all incoming network packets that make up one HTTP request in an instance of
+            // io.netty.buffer.CompositeByteBuf (think of it as a buffer of buffers). Once its capacity is reached, the buffer will iterate
+            // over its individual entries and put them into larger buffers (see io.netty.buffer.CompositeByteBuf#consolidateIfNeeded()
+            // for implementation details). We want to to resize that buffer because this leads to additional garbage on the heap and also
+            // increases the application's native memory footprint (as direct byte buffers hold their contents off-heap).
+            //
+            // With this setting we control the CompositeByteBuf's capacity (which is by default 1024, see
+            // io.netty.handler.codec.MessageAggregator#DEFAULT_MAX_COMPOSITEBUFFER_COMPONENTS). To determine a proper default capacity for
+            // that buffer, we need to consider that the upper bound for the size of HTTP requests is determined by `maxContentLength`. The
+            // number of buffers that are needed depend on how often Netty reads network packets which depends on the network type (MTU).
+            // We assume here that Elasticsearch receives HTTP requests via an Ethernet connection which has a MTU of 1500 bytes.
+            //
+            // Note that we are *not* pre-allocating any memory based on this setting but rather determine the CompositeByteBuf's capacity.
+            // The tradeoff is between less (but larger) buffers that are contained in the CompositeByteBuf and more (but smaller) buffers.
+            // With the default max content length of 100MB and a MTU of 1500 bytes we would allow 69905 entries.
```

Comment:
It is a nit, but would you please use the multi-line style:

```
/*
 *
 */
```?

Reply:
Sure. Addressed in https://github.com/elastic/elasticsearch/pull/29448/commits/b5c14fd7af00803293c47d69877c8991a083c789.